tfjs-backend-webgpu/src/backend_webgpu.ts

```diff
@@ -340,12 +338,83 @@ export class WebGPUBackend extends KernelBackend {
     return res;
   }
 
+  private getAndSaveProgram(
+      key: number, getProgram: () => webgpu_program.WebGPUProgram) {
+    if (!this.programCache.has(key)) {
+      this.programCache.set(key, getProgram());
+    }
+    return this.programCache.get(key);
+  }
+
+  // This hash algorithm came from https://source.chromium.org/chromium/chromium/src/+/master:third_party/dawn/src/common/HashUtils.h;l=47;bpv=1;bpt=0
```

Comment:
I'm not confident that porting this hash-combine function from C++ to JS is a good idea, because JS numbers behave so differently (in particular, `+` is a f64 operation, not a 2's-complement integer operation). Maybe there's an existing JS hash-combine function you can find?